tools/git_utils_test.py

```diff
@@ -0,0 +1,153 @@
+        test_file = self.tmp_path / "test_file.txt"
+        test_file.write_text("hello")
+
+        self._cmd(["git", "add", "test_file.txt"])
+        self._cmd(["git", "commit", "-m", "test_file"])
+        test_file_commit_hash = self._cmd(["git", "rev-parse", "HEAD"])
+
+        files_tracked = git_utils.ls_files(git_directory=self.tmp_path)
+        self.assertEqual(
+            files_tracked,
+            [
+                test_file.relative_to(self.tmp_path),
+            ],
+        )
+        self.assertEqual(
+            git_utils.get_files_changed_at_commit(
+                commit=test_file_commit_hash, git_directory=self.tmp_path
+            ),
+            [test_file.relative_to(self.tmp_path)],
+        )
+
+        # Check follow works as expected
+        self._cmd(["git", "mv", "test_file.txt", "moved_file.txt"])
+        self._cmd(["git", "commit", "-m", "moved_file"])
+        mv_commit_hash = self._cmd(["git", "rev-parse", "HEAD"])
+
+        self.assertEqual(
+            git_utils.list_file_commits("moved_file.txt", self.tmp_path),
+            # We're picking up its original
+            [mv_commit_hash, test_file_commit_hash],
+        )
+
+        self.assertEqual(
+            git_utils.get_commits(self.tmp_path),
+            [mv_commit_hash, test_file_commit_hash, first_hash],
+        )
+
+        # Show follow map is behaving as expected
+        follow_map = git_utils.get_file_commit_map_from_follow(self.tmp_path)
+        expected_follow = git_utils.FileCommitMap(
+            commit_map={
+                first_hash: set(),
+                test_file_commit_hash: set([pathlib.Path("moved_file.txt")]),
+                mv_commit_hash: set([pathlib.Path("moved_file.txt")]),
+            },
+            file_map={
+                pathlib.Path("moved_file.txt"): [
+                    mv_commit_hash,
+                    test_file_commit_hash,
+                ],
+            },
+        )
+        self.assertEqual(follow_map, expected_follow)
+
+        # Show how list misses out on tracking a file through history
+        list_map = git_utils.get_file_commit_map_from_list(self.tmp_path)
+        expected_list = git_utils.FileCommitMap(
+            commit_map={
+                first_hash: set(),
+                test_file_commit_hash: set([pathlib.Path("test_file.txt")]),
+                mv_commit_hash: set(
+                    [
+                        pathlib.Path("test_file.txt"),
+                        pathlib.Path("moved_file.txt"),
+                    ]
+                ),
+            },
+            file_map={
+                pathlib.Path("test_file.txt"): [
+                    mv_commit_hash,
+                    test_file_commit_hash,
+                ],
+                pathlib.Path("moved_file.txt"): [
+                    mv_commit_hash,
+                ],
+            },
+        )
+        self.assertEqual(list_map, expected_list)
+
+
+if __name__ == "__main__":
+    unittest.main()
```

Comment:
do you not exclude test files from coverage tests? :raised_eyebrow:

Reply:
nah, I just let it ride ;)

Reply:
gotta make sure you're actually running those bad boys